include/psa/crypto_values.h

```diff
@@ -892,6 +892,34 @@
 #define PSA_MAC_TRUNCATED_LENGTH(mac_alg)                               \
     (((mac_alg) & PSA_ALG_MAC_TRUNCATION_MASK) >> PSA_MAC_TRUNCATION_OFFSET)
 
+/* In the encoding of a MAC algorithm, the bit corresponding to
+ * PSA_ALG_MAC_MINIMUM_LENGTH_FLAG encodes the fact that the algorithm is
+ * a usage algorithm, which allows any algorithm corresponding to the same
+ * base class and a tag length greater or equal than the one encoded in
+ * PSA_ALG_MAC_TRUNCATION_MASK. */
+#define PSA_ALG_MAC_MINIMUM_LENGTH_FLAG         ((psa_algorithm_t)0x00008000)
+
+/** Macro to build a MAC minimum-tag-length usage algorithm.
+ *
+ * A mininimum-tag-length MAC usage algorithm contains all MAC algorithms
+ * sharing the same base algorithm, and where the tag length of the specific
+ * algorithm is equal to or larger then the usage's minimum tag length.
+ *
+ * \param mac_alg       A MAC algorithm identifier (value of type
+ *                      #psa_algorithm_t such that #PSA_ALG_IS_MAC(\p alg)
+ *                      is true).
+ * \param tag_length    Desired minimum length of the authentication tag in
+ *                      bytes.
+ *
+ * \return              The corresponding MAC usage algorithm with the
+ *                      specified minimum length.
+ * \return              Unspecified if \p alg is not a supported
+ *                      MAC algorithm or if \p tag_length is not valid
+ *                      for the specified MAC algorithm.
+ */
+#define PSA_ALG_MAC_WITH_MINIMUM_LENGTH_TAG(mac_alg, tag_length) \
```

Comment:
I find the name confusing: “with minimum length tag” makes me thing it builds something with the minimum valid length, not that it builds something where the specified length is a minimum. `PSA_ALG_MAC_WITH_MINIMUM_TAG_LENGTH` would be less confusing, although still potentially ambiguous. I'd prefer something like `PSA_ALG_MAC_WITH_AT_LEAST_LENGTH` or `PSA_ALG_TRUNCATED_MAC_AT_LEAST`.

Reply:
I took @athoelke 's recommended name. I propose you guys agree on a name first, after which I can update the PR accordingly.

Reply:
I have to admit that the recommendation was the result of an inconclusive and unsatisfactory search, and probably not enough consideration time, review and feedback.

Ideally we have an API that does not lend itself to avoidable-but-hard-to-spot errors. Consistency is very important, as is "does what it says". On the consistency front, MAC and AEAD have different naming conventions, because MAC is the name of the output as well as the algorithm category, whereas AEAD describes the algorithm but has a 'tag' as an output.

Naming options that better indicate that this is not a request for a minimum length MAC (or AEAD tag), and follow the MAC/AEAD-tag patterns include the following (as well as Gilles' suggestions above):

* `PSA_ALG_VARIABLE_LENGTH_MAC(mac_alg, min_len)` / `PSA_ALG_AEAD_WITH_VARIABLE_LENGTH_TAG(mac_alg, min_len)`
* `PSA_ALG_MULTI_LENGTH_MAC(mac_alg, min_len)` / `PSA_ALG_AEAD_WITH_MULTI_LENGTH_TAG(mac_alg, min_len)`
* `PSA_ALG_MAYBE_TRUNCATED_MAC(mac_alg, min_len)` / `PSA_ALG_AEAD_WITH_MAYBE_SHORTENED_TAG(mac_alg, min_len)` (🙄)
* `PSA_ALG_AT_LEAST_THIS_LENGTH_MAC(mac_alg, min_len)` / `PSA_ALG_AEAD_WITH_AT_LEAST_THIS_LENGTH_TAG(mac_alg, min_len)`

For clarity, we might want to ditch the MAC pattern and just duplicate the MAC term in the API, such as `PSA_ALG_MAC_WITH_VARIABLE_LENGTH_MAC(mac_alg, min_len)`

Votes, comments, and other suggestions are **VERY** welcome. ;-)

Reply:
1. I'd throw out multi-length, since that to me sounds like it contains a list of specified options
2. To me, `VARIABLE_LENGTH` sounds like it would allow anything allowed by the base algorithm. I wouldn't expect it to take an argument.
3. similarly, `MAYBE_TRUNCATED` just sounds too weird for a security library. The word 'maybe' shouldn't appear anywhere in critical code IMO.

I honestly don't think `MINIMUM_LENGTH` is such a bad/confusing description. It's important to note that minimum != minimal. I'd expect `XXX_WITH_MINIMAL_LENGTH` to give me the absolute lowest acceptable (=minimal) tag length, but `XXX_WITH_MINIMUM_LENGTH` a guarantee of at least a certain (=minimum) length.

If the linguistic argument isn't good enough, then I guess I'd be willing to settle for `PSA_ALG_AT_LEAST_THIS_LENGTH_MAC`/`PSA_ALG_AEAD_WITH_AT_LEAST_THIS_LENGTH_TAG`. It's definitely a mouthful, though.

Reply:
Still waiting for input here...

Reply:
“Minimum” _can_ mean “minimal”. I'd prefer an unambiguous name. I like “minimum” for expressing the concept, but I can't think of a concise name that doesn't have an obvious “minimal” interpretation.

I agree that “maybe” is bad. It isn't even accurate: CMAC-at-least-8 doesn't mean “maybe 8 or maybe untruncated” but “any length from 8 to 16”. I also agree that “multi” evokes a list rather than a range and so is not accurate either.

I'm lukewarm to “variable”. It's accurate, but I'm not convinced this is the distinguishing feature, as opposed to specifying a minimum length.

Reply:
I agree with almost all these comments on the naming options - but side with Gilles on the "minimum" vs "minimal" point. Arm tries not to depend on nuances of language in an API/spec that has to be understood by non-fluent readers of English.

It could be that `PSA_ALG_AT_LEAST_THIS_LENGTH_MAC`/`PSA_ALG_AEAD_WITH_AT_LEAST_THIS_LENGTH_TAG` are the least bad options we have on the table right now...

Reply:
Can we then move this along by agreeing on `PSA_ALG_AT_LEAST_THIS_LENGTH_MAC` / `PSA_ALG_AEAD_WITH_AT_LEAST_THIS_LENGTH_TAG`? @athoelke @gilles-peskine-arm ?

Reply:
Guys?

Reply:
I haven't been able to construct anything that is better on balance. @gilles-peskine-arm - can we live with this as an API identifier? - I am prepared to do so.

Reply:
I'm ok with `PSA_ALG_AT_LEAST_THIS_LENGTH_MAC` and `PSA_ALG_AEAD_WITH_AT_LEAST_THIS_LENGTH_TAG`. I find them ugly, but not misleading, unlike the other proposals in this thread which I find misleading.

Reply:
Done.